File: 143-reorder-list/reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;

        while(fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode head2 = slow.next;
        slow.next = null;

        slow = null;
        fast = head2;

        while(head2 != null) {
            fast = head2.next;
            head2.next = slow;
            slow = head2;
            head2 = fast;
        }

        head2 = slow;
        slow = head;

        while(slow != null && head2 != null) {
            fast = slow.next;
            slow.next = head2;
            slow = fast;
            fast = head2.next;
            head2.next = slow;
            head2 = fast;
        }

    }
}

File: 143-reorder-list/reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow = head
        fast = head
        prev = None

        while fast and fast.next:
            prev = slow
            slow = slow.next
            fast = fast.next.next

        if fast:
            head_rev = slow.next
            slow.next = None
        else:
            head_rev = slow
            prev.next = None

        if not head_rev:
            return

        prev = None
        fut = head_rev

        while fut.next:
            fut = fut.next
            head_rev.next = prev
            prev = head_rev
            head_rev = fut

        head_rev.next = prev

        odd = head
        fut = odd.next
        while fut and head_rev:
            odd.next = head_rev
            odd = odd.next
            head_rev = head_rev.next
            odd.next = fut
            odd = odd.next
            fut = fut.next

        if head_rev:
            odd.next = head_rev
